ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.hpp

```diff
@@ -0,0 +1,56 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <functional>
+#include <optional>
+
+#include "ttnn/tensor/tensor.hpp"
+#include "ttnn/operations/eltwise/unary/common/unary_op_types.hpp"
+#include "unary_program_factory.hpp"
+#include "unary_sharded_program_factory.hpp"
+
+#include "ttnn/device_operation.hpp"
+#include "ttnn/decorators.hpp"
+
+#include "unary_device_operation_types.hpp"
+
+
+namespace ttnn::operations::unary {
+
+struct UnaryDeviceOperation {
+
+    using operation_attributes_t = unary::operation_attributes_t;
+    using tensor_args_t = unary::tensor_args_t;
+    using shape_return_value_t = unary::shape_return_value_t;
+    using tensor_return_value_t = unary::tensor_return_value_t;
+    using program_factory_t = std::variant<program::UnaryProgramFactory, program::UnaryShardedProgramFactory>;
```

Comment:
This is controversial.

It does not achieve the goal is hiding programs from users of device operation (which I believe is desirable).
And it created another indirection for parameters.

But because programs and op types are not not nested, code of programs looks cleaner.

My recommendation here is to consider using a proper type erasure for a program to avoid variant. Our requirement to the underlying type
* can create a program
* can update runtime args

This should allow to hide underlying programs from consumers of device operation header.

Reply:
Wouldn't the user of device operation only care about `invoke()` and the inputs and outputs associated with it?

In terms of headers, usually those even include private fields and methods of classes, which can expose much more detailed internals.

Other ways to hide underlying:
- make these protected or private
- not have the prim operations in pybinds, limiting the scope of who consumes the headers